keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+  String string
+
+  // +default=[popcorn, chips]
+  StringList []string
+}
+```
+while still allowing more complex expressions:
+```
+// +default=[{port: 80, name: http}]
+```
+
+Forcing the "one-line" subset allows us to sidestep a lot of the YAML
+weirdness in syntax, but we can still just use a YAML parser to parse
+the expressions if we want.
+
+### Built-in Defaulting
+
+Because built-in types are serialized and deserialized as Go structures,
+there are semantics that are imposed on us. The two main ones are:
+1. We can't differentiate between null, unspecified and zero-value,
+2. non-pointer structs are always present (and hence have to be
+recursively defaulted), they also can't be omitEmpty'd.
+
+The first point is solved by the removed nulls described in the previous
+section (and invalid zero-values can't be present in objects), and the
+second point can be solved by forcing non-pointer structs to have a `{}`
+default. To address that, we're proposing that the `default` marker is
+forbidden on non-pointer structs (pointer structs can behave as
+expected) and the default field of the openapi is automatically set to
+`{}`. This only applies to OpenAPI generated from built-in types or from
+kubebuilder. CRDs can still have any defaulting they have for
+structures, but we strongly suggest that OpenAPI generated from Go types
+should follow that semantics.
+
+Computed defaults (through defaulting functions) will not be supported
+as a declarative marker and will continue being supported through the
+existing defaulting functions.
+
+### Examples
+
+#### Non-pointer structs
+
+Given the following type, defined either in Kubebuilder or as a built-in type:
+```golang
+type Root struct {
+  Entry SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      default: {}
+      type: object
+      properties:
+        name:
+          type: string
+          default: default-name
+        number:
+          type: integer
+          default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+>>> default('{entry: {name: "", number: 0}}') # empty string is different from null or unspecified
+{"entry": {"name": "", "number": 0}}
+```
+
+On the other hand, this is forbidden:
+```golang
+Type Root struct {
+  // Defaults on non-pointer structs are FORBIDDEN:
```

Comment:
Forbidden  or ignored?  What actually happens if someone specifies this?  Will CI fail?

Reply:
Forbidden. It will fail.